core/tee/tee_svc_cryp.c

```diff
@@ -2095,6 +2139,134 @@ tee_svc_obj_generate_key_x25519(struct tee_obj *o,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result
+tee_svc_obj_generate_key_ed25519(struct tee_obj *o,
+				 const struct tee_cryp_obj_type_props
+							*type_props,
+				 uint32_t key_size,
+				 const TEE_Attribute *params,
+				 uint32_t param_count)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct ed25519_keypair *key = NULL;
+
+	/* Copy the present attributes into the obj before starting */
+	res = tee_svc_cryp_obj_populate_type(o, type_props, params,
+					     param_count);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	key = o->attr;
+
+	res = crypto_acipher_gen_ed25519_key(key, key_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/* Set bits for the generated public and private key */
+	set_attribute(o, type_props, TEE_ATTR_ED25519_PRIVATE_VALUE);
+	set_attribute(o, type_props, TEE_ATTR_ED25519_PUBLIC_VALUE);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+tee_svc_obj_ed25519_sign(struct ed25519_keypair *key,
+			 const uint8_t *msg, size_t msg_len,
+			 uint8_t *sig, size_t *sig_len,
+			 const TEE_Attribute *params, size_t num_params)
+{
+	TEE_Result err;
+	size_t n;
+	size_t ctx_len = 0;
+	uint8_t *ctx = NULL;
+	bool ph_flag = false;
+	bool cx_flag = false;
+
+	for (n = 0; n < num_params; n++) {
+		switch (params[n].attributeID) {
+		case TEE_ATTR_EDDSA_PREHASH:
+			ph_flag = true;
+			break;
+
+		case TEE_ATTR_EDDSA_CTX:
+			/* only first provided context if effective */
+			if (cx_flag)
+				break;
+			ctx_len = params[n].content.ref.length;
+			if (ctx_len > 255)
+				return TEE_ERROR_BAD_PARAMETERS;
+			ctx = mempool_calloc(mempool_default, 1, ctx_len + 1);
+			if (!ctx)
+				return TEE_ERROR_OUT_OF_MEMORY;
+			memcpy(ctx, params[n].content.ref.buffer, ctx_len);
+			ctx[ctx_len] = 0;
+			cx_flag = true;
+			break;
+
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	if (ph_flag || cx_flag) {
+		err = crypto_acipher_ed25519ctx_sign(key, msg, msg_len, sig,
+						     sig_len, ph_flag,
+						     ctx, ctx_len);
+		if (ctx)
+			mempool_free(mempool_default, ctx);
+		return err;
+	}
+
+	return crypto_acipher_ed25519_sign(key, msg, msg_len, sig, sig_len);
+}
+
+static TEE_Result
+tee_svc_obj_ed25519_verify(struct ed25519_keypair *key,
+			   const uint8_t *msg, size_t msg_len,
+			   const uint8_t *sig, size_t sig_len,
+			   const TEE_Attribute *params, size_t num_params)
+{
+	TEE_Result err;
+	size_t n;
```

Comment:
ditto

Reply:
fixed in [785b4be](https://github.com/OP-TEE/optee_os/pull/5486/commits/785b4be2ede4583aa7e251851bdb6ed7905ccbc3)